### PR TITLE
Adding Sub-channel Scaling Grouped GEMMs

### PR DESCRIPTION
**Pull request by Shreya-gaur**
MR: Subchannel-scaled Grouped GEMM kernels (basic_gemm + dgrad + wgrad) for cuDNN Frontend

**Changes**
- sggemm integrated — Adds the initial grouped_gemm_unfused_subchannel_scaled kernel under python/cudnn/gemm/cutedsl/grouped/unfused_subchannel_scaled/: Blackwell + Rubin kernel implementations, API surface, moe_kernel_helpers.py/moe_sched_extension.py scheduling support, docs, and test/dispatch wiring. This is the foundational commit establishing the grouped/ subchannel-scaled kernel package layout reused by later kernels.
- sggemm + dswiglu + dquant kernel integrated — Adds grouped_gemm_dswiglu_subchannel_scaled: Blackwell kernel, API, tests/utils, and extends the shared moe_kernel_helpers.py/moe_sched_extension.py to support dswiglu/dquant scheduling; minor follow-on fix to the unfused kernel.
- sggemm wgrad kernel integrated — Adds grouped_gemm_wgrad_subchannel_scaled (weight-gradient grouped GEMM) under .../wgrad_subchannel_scaled/, including Blackwell + Rubin kernel implementations, API, shared helper/scheduling extensions, docs, and kernel-dispatch test wiring.
- numerical tests for e5m3 in wgrad and unfused kernels — Extends the wgrad and unfused subchannel-scaled test suites with e5m3 numerical coverage, plus corresponding doc updates.
- rubin version of dswiglu sggemm kernel — Adds the Rubin implementation of grouped_gemm_dswiglu_subchannel_scaled (~5.3K LOC), updates the dswiglu API and tests/utils accordingly, and registers it in the Rubin kernel dispatch test.
- Cleaning up dead code — Removes ~1.5K lines of dead/unused code across the dswiglu, unfused, and wgrad subchannel-scaled kernels (both Blackwell and Rubin variants) and their APIs/tests ([[Subchannel dead-code pass]]).
- Changing the comments — Comment-only cleanup pass on the dswiglu subchannel-scaled kernel (Blackwell + Rubin), reducing verbosity (~565 lines net removed) without functional changes.

**Test plan**

- [x] test_grouped_gemm_unfused_subchannel_scaled.py passes on Blackwell and Rubin
- [x] test_grouped_gemm_dswiglu_subchannel_scaled.py passes on Blackwell and Rubin
- [x] test_grouped_gemm_wgrad_subchannel_scaled.py passes on Blackwell and Rubin
- [x] test_rubin_kernel_dispatch.py covers all Rubin kernel registrations (unfused, dswiglu, wgrad)
- [x] e5m3 numerical tests pass for wgrad and unfused kernel

## Before submitting

- [ ] I agree to license this contribution under the terms of [LICENSE.txt](https://github.com/NVIDIA/cudnn-frontend/blob/develop/LICENSE.txt).
- [ ] I ran `pre-commit run` and committed any formatting changes.
- [ ] I reviewed the Hard Rules in the `AGENTS.md` for each directory this PR touches (see [root AGENTS.md § Reviewing a PR](https://github.com/NVIDIA/cudnn-frontend/blob/develop/AGENTS.md#reviewing-a-pr-human-or-agent)) and my changes comply, or I explain the exception below.
- [ ] I added GitHub labels: one `cat-*`, one or more `area:*` / `op:*`, and one `orig-*` (see [label list](https://github.com/NVIDIA/cudnn-frontend/labels)).

### CODE REVIEW

File: python/cudnn/gemm/cutedsl/grouped/__init__.py [hunk 1]
```diff
@@ -79,6 +83,8 @@
     "grouped_gemm_wgrad_wrapper_sm100",
     "GroupedGemmSm100",
     "grouped_gemm_wrapper_sm100",
+    "GroupedGemmUnfusedSubchannelScaledSm100",
```
Comment (Anerudhan) [inline python/cudnn/gemm/cutedsl/grouped/__init__.py:94]: Isnt this sm107 kernel?

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Codex bot review · model: `gpt-6-astra`**

Requesting OSS CI for the current subchannel-scaled GEMM changes (`fa1de1d6f481`). Local SM100 tests passed, and review is checking additional API boundaries.

For future kernel updates, use the command above to request OSS CI after pushing the revision you want tested. If your account cannot trigger it, ask a maintainer to issue the request.

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model: `gpt-6-astra`**

The first request was rejected because this trigger treats trailing explanatory text as target names; a signed comment with the command at the end was not acknowledged either. I am issuing the standalone command immediately below on behalf of this review bot. For future requests, use a comment containing only `@cudnn-ci-bot run oss`; ask a maintainer if your account lacks trigger permissions.

**Comment by YangXu1990uiuc** [comment]
@cudnn-ci-bot run oss

**Review by YangXu1990uiuc — APPROVED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `fa1de1d6f481` against base `8021591e84c2`.

The two-level scaling design and separate dense/discrete APIs make sense. The three new API suites pass on both SM100 and SM107, but the additional stream, device-cache and CUDA Graph probes below expose API integration issues. No P0 found; the P1/P2 items remain for the owner to address.

**[P1] Initialize atomic outputs on the supplied execution stream** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/fa1de1d6f481c833d67b555ded16f6ec86f9cb07/python/cudnn/gemm/cutedsl/grouped/dswiglu_subchannel_scaled/api.py#L1239))

When current_stream differs from the ambient PyTorch stream, these torch.zeros calls run on the ambient stream while api.execute launches on current_stream. The dprob/dbias accumulations and sfd2 atomic maxima therefore race their zero initialization. After warming the kernel and synchronizing its inputs, delaying only the ambient stream produced all-zero dprob, dbias and sfd2 where the reference was nonzero. Please put the initialization and stream-dependent preparation under the existing _torch_stream_context(current_stream, device), and cover a delayed ambient stream.

Evidence: SM100: M/expert=256, N=256, K=512, 2 experts, BF16 D and FP32 dbias. Delaying the ambient stream reproduced the failure; enclosing the same call in _torch_stream_context made all three outputs byte-exact against the warm reference. Attribution: Introduced in the new dSwiGLU wrapper; shared by its architecture branches.

**[P1] Include the CUDA device in the unfused and dSwiGLU cache keys** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/fa1de1d6f481c833d67b555ded16f6ec86f9cb07/python/cudnn/gemm/cutedsl/grouped/unfused_subchannel_scaled/api.py#L1121))

device_type is only blackwell/rubin, and the tensor signatures omit tensor.device. A same-shape call on a second GPU consequently reuses the first GPU's API object, including its workspace and device-bound descriptors. This loses per-device scheduler/tensormap scratch isolation and can use foreign-device storage or race an in-flight call on the first GPU. The dSwiGLU cache has the same omission at its common key construction. Please key by the actual input device/context and ensure device consistency; the existing wgrad tensor signature already includes device identity.

Evidence: A two-SM100-GPU probe warmed each wrapper on cuda:0, then invoked matching inputs on cuda:1. Both caches stayed at one entry and selected an object with workspace_device=cuda:0 and descriptor_device=cuda:0 for input_device=cuda:1. The second launch was intercepted: this proves cache selection, not an executed cross-device kernel failure. Attribution: Introduced in the two new wrapper caches.

**[P2] Prepare discrete output pointer tables outside the captured execute path** ([source](https://github.com/NVIDIA/cudnn-frontend/blob/fa1de1d6f481c833d67b555ded16f6ec86f9cb07/python/cudnn/gemm/cutedsl/grouped/wgrad_subchannel_scaled/api.py#L597))

With output_mode="discrete" and a preallocated wgrad_tensor, execute still builds a Python pointer list and copies it to a new CUDA tensor on every call. A warmed invocation inside torch.cuda.graph fails here with "Cannot copy between CPU and CUDA tensors during CUDA graph capture unless the CPU tensor is pinned." This also adds allocation/copy work to the steady execute path. Please prepare/bind the table outside execute, or require and document precomputed wgrad_ptrs for capture; avoid a per-call host transfer in the prepared path.

Evidence: SM100: hidden/intermediate=128, two experts with 512 tokens each, sgk=512. Capturing the warmed wrapper with a preallocated output failed at this line; supplying the precomputed device wgrad_ptrs made capture/replay pass correctness. Attribution: Introduced in the new API's discrete wgrad_tensor convenience path.

- **[P2] Still open:** [Resolve the already-reported incompatible dSwiGLU defaults](https://github.com/NVIDIA/cudnn-frontend/pull/953#discussion_r3973564139).

Validation: Explicitly ran test_grouped_gemm_{unfused,dswiglu,wgrad}_subchannel_scaled.py with -m "L0 or L1": SM100 57 passed / 26 skipped; SM107 83 passed. Head fa1de1d6f481; PyTorch 2.13.0+cu130, cuDNN 9.25.1, CuTe DSL 4.8 development wheel. Verified the checkout import and target GPU on each machine. Additional SM100 negative/control probes confirmed the stream and graph-capture findings; two-GPU interception confirmed the cache-device collision. Inspected API contracts, architecture dispatch, second-level scale handling and the shared scheduler/helper changes. Existing helper defaults retain their prior behavior. Publication recheck against develop 8021591e84c2: unchanged merge-base and no changes to grouped CuTe GEMM, api_base or tensor_adapter since the originally captured base. The intervening datatype change adds DLPack FP4 export helpers without changing the dtype conversion used here. New discussion consists of the CI request/acknowledgments, with no new owner fix or conflicting review evidence.

Limitations: OSS CI is running for fa1de1d, as confirmed by the CI bot. Its outcome is pending; local targeted tests do not establish full-suite CI success. No baseline/head timing claim: these are new subchannel operations without a like-for-like baseline API. The host pointer-table work above was verified functionally; its timing impact was not measured.

Approved under the trial's no-P0 threshold; any P1/P2 findings above remain for the owner to address. Merge timing stays with the owner.

**Comment by YangXu1990uiuc** [comment]
**Codex bot review · model: `gpt-6-astra`**

CI follow-up for `fa1de1d6f481`: the requested OSS test lanes completed successfully. The remaining failed check is the comment-word scan at `python/cudnn/gemm/cutedsl/grouped/dswiglu_subchannel_scaled/moe_blockscaled_grouped_gemm_dswiglu_subchannel_scaled_rubin.py:3491`, in the AST/constexpr explanatory comment. Rewording that comment should address this particular check; the correctness/API findings in the review remain separate.

**Comment by YangXu1990uiuc** [comment]
**Codex bot review follow-up · model: `gpt-6-astra`**

Please make sure both P1s in [the review](https://github.com/NVIDIA/cudnn-frontend/pull/953#pullrequestreview-5202798398) are addressed before merging `fa1de1d6f481`: the explicit-stream initialization race can silently corrupt `dprob`, `dbias` and `sfd2`, and the device-independent cache keys can reuse GPU-owned scratch on another device. These are high-risk correctness issues. Please include a delayed-ambient-stream regression and a same-shape, second-device check with the fixes.

File: python/cudnn/gemm/cutedsl/grouped/__init__.py (hunk 1 above)
Comment (Shreya-gaur) [reply python/cudnn/gemm/cutedsl/grouped/__init__.py]: The API base naming convention is the same as other kernels like under "quant" folder. In the API name the Sm100 is denoting SM100+ architectures. The kernels themselves are named as per the architecture chose by the API. If you see the class name in the kernel file, we explicitly name the rubin kernel with "Sm107"

**Review by YangXu1990uiuc — COMMENTED** [review]
**Codex bot review · model: `gpt-6-astra`**

Reviewed `4ef320ada431` against base `d34a6909c9d1`.

This update merges develop but leaves the previously reviewed subchannel APIs and kernels unchanged. The two high-risk P1 findings remain unresolved; please address them before merging. On this exact head I reproduced the explicit-stream initialization race again: dprob/dbias/sfd2 were all zero after delaying only the ambient stream, while their warm references had 512/1024/738 nonzero elements. Running the same call with initialization ordered on the supplied stream made all three outputs byte-exact against the reference. Existing approval is preserved under the agreed policy; this is a COMMENT, not another approval or a request to dismiss it. No broad CI was requested while these correctness issues remain.

- **[P1 · high risk] Still open:** [Order atomic-output initialization on the supplied execution stream](https://github.com/NVIDIA/cudnn-frontend/pull/953#pullrequestreview-5202798398).

- **[P1 · high risk] Still open:** [Keep cached workspace and descriptors isolated by CUDA device](https://github.com/NVIDIA/cudnn-frontend/pull/953#pullrequestreview-5202798398).

- **[P2] Still open:** [Prepare discrete output pointer tables before captured execution](https://github.com/NVIDIA/cudnn-frontend/pull/953#pullrequestreview-5202798398).

- **[P2] Still open:** [Make the default output dtype and layout compatible](https://github.com/NVIDIA/cudnn-frontend/pull/953#discussion_r3973564139).

Validation: Current-head local SM100 stream reproducer and ordered control, M/expert=256, N=256, K=512, two experts, BF16 D and FP32 dbias. Inputs were synchronized before delaying the ambient stream. Incorrect-call maximum errors were 0.005095 / 1.926162 / 0.00022175 for dprob/dbias/sfd2; the ordered control had zero error for all three. Verified the current source import and native-extension compatibility. Compared the subchannel APIs, kernels and tests with fa1de1d6f481; they are unchanged. Rechecked affected contracts against current develop d34a6909c9d1; its changes since the earlier captured base do not alter these grouped-GEMM paths. Prior broader evidence was 57 passed / 26 skipped on SM100 and 83 passed on SM107 at fa1de1d6f481. These are historical results, not new full-suite results for this merge. The cache-device selection and discrete-capture probes remain linked in the original review.

Limitations: No new full GPU suite, Rubin run, two-GPU launch, or performance comparison for this integration-only follow-up. The supported-call correctness blocker already reproduces on the current head. Earlier CI belongs to fa1de1d6f481 and is not substituted for current-head CI. Automatic broad CI is deferred under the maturity gate.

Comment only: this PR already has an active approval; preserving existing approvals.

Please make sure the high-risk P1 findings above are addressed before merging. Existing approvals are left unchanged.
